### PR TITLE
Add CSV support for batch output

### PR DESCRIPTION
**Pull request by mdsrosa**
This PR is a follow up on the PR https://github.com/dbcli/mycli/pull/270. This PR add CSV support for batch output feature.

```
$ cat query.txt | mycli -u root test_db --csv
id,email,name
1,email1@gmail.com,Foo Bar
2,emal2@gmail.com,Bar Foo
3,null,John Doe
```

Thanks @gubbin!

### CODE REVIEW

**Comment by mdsrosa** [comment]
Working on the tests right now.

**Comment by mdsrosa** [comment]
Okay. Tests fixed. 

Could you please review it? @amjith @tsroten

File: tests/test_main.py [hunk 1]
```diff
@@ -88,6 +88,22 @@ def test_batch_mode_table(executor):
...
+    run(executor, '''create table test(a text, b text)''')
+    run(executor, '''insert into test (a, b) values('abc', 'def'), ('ghi', 'jkl')''')
+
+    sql = 'select * from test;'
+
+    runner = CliRunner()
+    result = runner.invoke(cli, args=CLI_ARGS + ['--csv'], input=sql)
+
+    expected = 'a,b\nabc,def\nghi,jkl\n'
+    result_output = result.output.replace('\x00', '') # python 3
```
Comment (mdsrosa) [inline tests/test_main.py]: @amjith @tsroten I'm not very happy with this "hacky" solution for the test. Do you have any idea on this?

Comment (amjith) [inline tests/test_main.py]: I'm not sure why the `\x00` is introduced. I wonder if this is a bug in click's CliRunner.

File: mycli/main.py [hunk 2]
```diff
@@ -832,7 +856,23 @@ def format_output(title, cur, headers, status, table_format, expanded=False, max
         headers = [utf8tounicode(x) for x in headers]
         if expanded:
             output.append(expanded_table(cur, headers))
-        elif table_format is not None:
+        elif table_format is None:
+            output.append('\t'.join(headers))
+            for row in cur:
+                output.append('\t'.join([str(r) for r in row]))
```
Comment (tsroten) [inline mycli/main.py]: I think you can use our tabulate library for this. It has a `tsv` format. It looks like if you were to set `table_format` to `'tsv'` if it is `None`, then the `else:` statement below will take care of this code.

File: mycli/main.py [hunk 3]
```diff
@@ -13,6 +14,12 @@
 from random import choice
 from io import open
 
+# support StringIO for Python 2 and 3
+try:
+    from StringIO import StringIO
```
Comment (tsroten) [inline mycli/main.py]: For Python2, we can get better performance if we do `from cStringIO import StringIO`.
https://docs.python.org/2.6/library/stringio.html#module-cStringIO

**Comment by tsroten** [comment]
@mdsrosa I'm glad we're finally getting CSV support!!

For the actual CSV generation -- I think the various truncate/getvalue calls are leading to an exponential amount of data in memory (it keeps overwriting previous rows with null bytes, which still take up memory). Instead, we could get by with a single `output.append(content.getvalue())` after the loop that appends *both* header and data to the output. Then we can drop the `truncate` commands.

File: mycli/main.py [hunk 4]
```diff
@@ -818,7 +827,22 @@ def cli(database, user, host, port, socket, password, dbname,
                 confirm_destructive_query(stdin_text) is False):
             exit(0)
         try:
-            mycli.run_query(stdin_text, table_format=table)
```
Comment (tsroten) [inline mycli/main.py]: We now have two sets of code for running a query -- one that continues to use `run_query` and one that uses this version. It looks like the big difference is the newline handling. Is there any way we could continue using the `run_query` function, but also remove the extra newline? Maybe a helper function that looks at the settings to determine if we should output a newline or something?

**Comment by mdsrosa** [comment]
```
@mdsrosa I'm glad we're finally getting CSV support!!

For the actual CSV generation -- I think the various truncate/getvalue calls are leading to an exponential amount of data in memory (it keeps overwriting previous rows with null bytes, which still take up memory). Instead, we could get by with a single output.append(content.getvalue()) after the loop that appends both header and data to the output. Then we can drop the truncate commands.
```

@tsroten Oh didn't know that it is how it behaves. Thank you for that knowlodge. I will try removing those extra calls to `truncate` and `getvalue()`.

File: mycli/main.py (hunk 4 above)
Comment (mdsrosa) [inline mycli/main.py]: @tsroten You're right. I'll refactor this out. Maybe adding a new parameter called `new_line` to the `run_query` function?

File: tests/test_main.py (hunk 1 above)
Comment (mdsrosa) [inline tests/test_main.py]: @tsroten You were right. I removed the `truncate` calls and the `\x00` took vacation. 😄 
Thanks, man!

File: mycli/main.py (hunk 4 above)
Comment (mdsrosa) [inline mycli/main.py]: Done.

File: mycli/main.py (hunk 2 above)
Comment (mdsrosa) [inline mycli/main.py]: @tsroten Yeah, turns out you're were on this one too. I've set the default value as `tsv` for the `table_format`. Thanks, man!

**Comment by tsroten** [comment]
Looking good! I think that this might have affected the `--execute` command's output. When I run it, it doesn't use my `table_format`.

**Comment by mdsrosa** [comment]
@tsroten I think you found a bug and I think I know the fix for it.

https://github.com/dbcli/mycli/blob/124ca067713a4896a398e25e2367989fbc192cc7/mycli/main.py#L808

```
if execute:
        try:
            mycli.run_query(execute, table_format=table)
```

Here it must be `mycli.run_query(execute, table_format=mycli.table_format`.  Correct?

**Comment by tsroten** [comment]
@mdsrosa I think that's the right spot. I believe that would ignore the absence of the `--table` argument then. It should only use `mycli.table_format` if the `table` argument is True. I think that's why the previous code passed in `table` -- then if `table` was true it used `mycli.table_format`.

**Comment by mdsrosa** [comment]
@tsroten Okay. So I should add a check for `--table` and if it's `True` I'll use `mycli.table_format` otherwise `table_format=None`.

**Comment by tsroten** [comment]
If we move it down to the `csv` logic we risk running into problems with the `isatty` and `stdin` code. I wonder if we could just do:
```python
if execute:
    try:
      mycli.run_query(execute, table_format=table_format if table else None)
```

**Comment by mdsrosa** [comment]
@tsroten Hmm. Sorry, could you please elaborate on the risks? I'm asking this so I can learn about more them. By the way, I ran the tests and they're passing. 😄

**Comment by tsroten** [comment]
It looks like the way you did it was fine. :)

My comment about it being risky was referring to bringing the `--execute` code down to around line 843 where the CSV code was. I didn't realize you meant you were going to also add CSV support to the `--execute` command.

*If* the `--execute` code was brought down below the `isatty` check, then in most scenarios, you could never use `--execute` because mycli would run like normal. But, I think you know that since you were the one to add the `--execute` code in the correct spot to begin with ;)

**Comment by mdsrosa** [comment]
@tsroten Oh, sorry for being totally clear about what I meant to do there.
So yeah, the `--execute` command is safe. Do you think this is ready to merge? 🙂

**Pull request merged**